File: 17.Longest_Substring_Without_Repeating_Characters/sol.go
package main

import (
	"fmt"
	"os"
	"runtime"
	"strings"
)

// ======================================================================
// 🧠 CHALLENGE: Longest Substring Without Repeating Characters (Go)
// ======================================================================
// Description:
// Given a string s, find the length of the longest substring without
// duplicate characters.
//
// 📋 Rules:
// 1. Substring must be contiguous.
// 2. No repeated characters inside the chosen substring.
// 3. Return the maximum length.
//
// Constraints:
// - 0 <= len(s) <= 5 * 10^4
// - s consists of English letters, digits, symbols, and spaces.
// ======================================================================

// #region [📚 Reference Solutions] (Solutions hidden as requested)
func lengthOfLongestSubstring1(s string) int {
	// Method: Sliding Window + Last Seen Index (Hash Map)
	// Complexity: Time O(N) | Space O(min(N, charset))

	last := make(map[byte]int, 128) // byte -> last index
	for i := 0; i < 128; i++ {
		last[byte(i)] = -1
	}

	l := 0
	best := 0

	for r := 0; r < len(s); r++ {
		ch := s[r]
		if last[ch] >= l {
			l = last[ch] + 1
		}
		last[ch] = r

		if r-l+1 > best {
			best = r - l + 1
		}
	}
	return best
}

// #endregion

// ======================================================================
//
//	#region [✍️ Practice Area]
//	Please write your solution between the markers below.
//
// ======================================================================
// <PRACTICE_START>
func LengthOfLongestSubstring(s string) int {
	// TODO: Implement your solution here.
	return 0
}
// <PRACTICE_END>
// #endregion

// ======================================================================
//  #region [🚀 Test Runner & Auto-Reset] (Do not modify below this line)
// ======================================================================

type TestCase struct {
	s        string
	expected int
}

func main() {
	runTests()
}

func runTests() {
	testCases := []TestCase{
		{"abcabcbb", 3},
		{"bbbbb", 1},
		{"pwwkew", 3},
		{"", 0},
		{" ", 1},
		{"dvdf", 3},    // "vdf"
		{"abba", 2},    // "ab" or "ba"
		{"tmmzuxt", 5}, // "mzuxt"
	}

	fmt.Printf("\n🧪 Testing your [LengthOfLongestSubstring] function...\n\n")

	header := fmt.Sprintf("%-25s | %-10s | %-10s | %s", "Input s", "Expected", "Actual", "Status")
	fmt.Println(header)
	fmt.Println(strings.Repeat("-", len(header)))

	allPass := true

	for _, tc := range testCases {
		actual := -999999
		statusIcon := "✅ PASS"

		func() {
			defer func() {
				if r := recover(); r != nil {
					actual = -999999
					statusIcon = "❌ FAIL"
					allPass = false
				}
			}()
			actual = LengthOfLongestSubstring(tc.s)
		}()

		if actual != tc.expected {
			statusIcon = "❌ FAIL"
			allPass = false
		}

		inStr := fmt.Sprintf("%q", tc.s)
		if len(inStr) > 23 {
			inStr = inStr[:20] + "..."
		}
		fmt.Printf("%-25s | %-10d | %-10d | %s\n", inStr, tc.expected, actual, statusIcon)
	}

	fmt.Println(strings.Repeat("-", len(header)))

	if allPass {
		fmt.Println("\n🎉 Fantastic! All test cases passed.")
		resetPracticeArea()
	} else {
		fmt.Println("\n⚠️  Some tests failed. Keep trying! (The file will not reset yet)")
	}
}

func resetPracticeArea() {
	fmt.Println("\n🔄 Resetting Practice Area to default state...")

	markerStart := "// <PRACTICE_" + "START>"
	markerEnd := "// <PRACTICE_" + "END>"

	defaultCode := []string{
		"func LengthOfLongestSubstring(s string) int {",
		"\t// TODO: Implement your solution here.",
		"\treturn 0",
		"}",
	}

	_, currentFile, _, ok := runtime.Caller(0)
	if !ok {
		fmt.Println("⚠️ Error: Could not determine file path. Reset cancelled.")
		return
	}

	content, err := os.ReadFile(currentFile)
	if err != nil {
		fmt.Printf("⚠️ Error reading file: %v\n", err)
		return
	}

	lines := strings.Split(string(content), "\n")
	startIdx := -1
	endIdx := -1

	for i, line := range lines {
		if strings.Contains(line, markerStart) {
			startIdx = i
		} else if strings.Contains(line, markerEnd) {
			endIdx = i
		}
	}

	if startIdx == -1 || endIdx == -1 || startIdx >= endIdx {
		fmt.Println("⚠️ Error: Markers not found or invalid. Reset cancelled.")
		return
	}

	newLines := make([]string, 0)
	newLines = append(newLines, lines[:startIdx+1]...)
	newLines = append(newLines, defaultCode...)
	newLines = append(newLines, lines[endIdx:]...)

	output := strings.Join(newLines, "\n")
	err = os.WriteFile(currentFile, []byte(output), 0644)
	if err != nil {
		fmt.Printf("⚠️ Error writing file: %v\n", err)
		return
	}

	fmt.Println("✨ Reset complete! The file is ready for a fresh start.")
}

// #endregion

File: 17.Longest_Substring_Without_Repeating_Characters/sol.py
import os
import sys
from typing import Dict

# ======================================================================
# 🧠 CHALLENGE: Longest Substring Without Repeating Characters (Python)
# ======================================================================
# Description:
# Given a string s, find the length of the longest substring without
# duplicate characters.
#
# 📋 Rules:
# 1. Substring must be contiguous.
# 2. No repeated characters inside the chosen substring.
# 3. Return the maximum length.
#
# 💡 Examples:
# - Input: "abcabcbb" => Output: 3  ("abc")
# - Input: "bbbbb"    => Output: 1  ("b")
# - Input: "pwwkew"   => Output: 3  ("wke")
#
# Constraints:
# - 0 <= len(s) <= 5 * 10^4
# - s consists of English letters, digits, symbols, and spaces.
# ======================================================================


# region [📚 Reference Solutions] (Solutions hidden as requested)
def length_of_longest_substring1(s: str) -> int:
    """
    Method: Sliding Window + Last Seen Index (Hash Map)
    Complexity: Time O(N) | Space O(min(N, charset))
    """
    last = {}  # char -> last index
    l = 0
    best = 0

    for r, ch in enumerate(s):
        if ch in last and last[ch] >= l:
            l = last[ch] + 1
        last[ch] = r

        win_len = r - l + 1
        if win_len > best:
            best = win_len

    return best


# endregion


# ======================================================================
#  region [✍️ Practice Area]
#  Please write your solution between the markers below.
# ======================================================================
# <PRACTICE_START>
def length_of_longest_substring(s: str) -> int:
    last = {}
    l = 0
    res = 0

    for r, ch in enumerate(s):
        if ch in last and last[ch] >= l:
            l = last[ch] + 1

        last[ch] = r

        res = max(res, r - l + 1)

    return res


# <PRACTICE_END>
# endregion

# ======================================================================
#  region [🚀 Test Runner & Auto-Reset] (Do not modify below this line)
# ======================================================================


def reset_practice_area():
    print("\n🔄 Resetting Practice Area to default state...")
    MARKER_START = "# <PRACTICE_" + "START>"
    MARKER_END = "# <PRACTICE_" + "END>"

    default_code_lines = [
        "def length_of_longest_substring(s: str) -> int:\n",
        "    # TODO: Implement your solution here.\n",
        "    return 0\n",
    ]

    try:
        current_file = __file__
        with open(current_file, "r", encoding="utf-8") as f:
            lines = f.readlines()

        start_idx = -1
        end_idx = -1

        for i, line in enumerate(lines):
            if MARKER_START in line:
                start_idx = i
            if MARKER_END in line:
                end_idx = i

        if start_idx == -1 or end_idx == -1 or start_idx >= end_idx:
            print("⚠️ Error: Markers not found. Reset cancelled.")
            return

        new_content = lines[: start_idx + 1] + default_code_lines + lines[end_idx:]

        with open(current_file, "w", encoding="utf-8") as f:
            f.writelines(new_content)

        print("✨ Reset complete! The file is ready for a fresh start.")
    except Exception as e:
        print(f"⚠️ Reset failed: {e}")


if __name__ == "__main__":
    test_cases = [
        ("abcabcbb", 3),
        ("bbbbb", 1),
        ("pwwkew", 3),
        ("", 0),
        (" ", 1),
        ("dvdf", 3),  # "vdf"
        ("abba", 2),  # "ab" or "ba"
        ("tmmzuxt", 5),  # "mzuxt"
    ]

    print(f"\n🧪 Testing your [length_of_longest_substring] function...\n")

    header = f"{'Input s':<25} | {'Expected':<10} | {'Actual':<10} | Status"
    print(header)
    print("-" * len(header))

    all_pass = True
    for s, expected in test_cases:
        try:
            actual = length_of_longest_substring(s)
        except Exception:
            actual = "Error"

        is_match = actual == expected
        status_icon = "✅ PASS" if is_match else "❌ FAIL"
        if not is_match:
            all_pass = False

        in_str = repr(s)
        if len(in_str) > 23:
            in_str = in_str[:20] + "..."
        print(f"{in_str:<25} | {str(expected):<10} | {str(actual):<10} | {status_icon}")

    print("-" * len(header))
    if all_pass:
        print("\n🎉 Fantastic! All test cases passed.")
        reset_practice_area()
    else:
        print("\n⚠️  Some tests failed. Keep trying! (The file will not reset yet)")
# endregion
